maven-core/src/main/java/org/apache/maven/utils/dag/CycleDetector.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.utils.dag;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author <a href="michal.maczka@dimatics.com">Michal Maczka</a>
+ *
+ */
+public class CycleDetector {
+
+    private static final Integer NOT_VISITED = 0;
+
+    private static final Integer VISITING = 1;
+
+    private static final Integer VISITED = 2;
+
+    public static List<String> hasCycle(final DAG graph) {
+        final List<Vertex> vertices = graph.getVertices();
+
+        final Map<Vertex, Integer> vertexStateMap = new HashMap<>();
+
+        List<String> retValue = null;
+
+        for (Vertex vertex : vertices) {
+            if (isNotVisited(vertex, vertexStateMap)) {
+                retValue = introducesCycle(vertex, vertexStateMap);
+
+                if (retValue != null) {
+                    break;
+                }
+            }
+        }
+
+        return retValue;
+    }
+
+    /**
+     * This method will be called when an edge leading to given vertex was added and we want to check if introduction of
```

Comment:
is called
is added
i.e. present tense always in tech writing

Reply:
Done.